faculty/clients/api.py

```diff
@@ -0,0 +1,635 @@
+    id = fields.UUID(data_key="keyId", required=True)
+    material = fields.String(required=True)
+    # Development instance keys do not have these fields
+    enabled = fields.Boolean(load_default=None)
+    label = fields.String(load_default=None)
+
+    @post_load
+    def make_apikey(self, data, **kwargs):
+        return APIKey(**data)
+
+
+class RemainingResourceLimitSchema(BaseSchema):
+    allowed_max_milli_cpus = fields.Integer(
+        data_key="allowedMaxMilliCpus", required=True
+    )
+    allowed_max_memory_mb = fields.Integer(
+        data_key="allowedMaxMemoryMb", required=True
+    )
+    remaining_milli_cpus = fields.Integer(
+        data_key="remainingMilliCpus", required=True
+    )
+    remaining_memory_mb = fields.Integer(
+        data_key="remaininigMemoryMb", required=True
+    )
+
+
+class DevInstanceIdSchema(BaseSchema):
+    instance_id = fields.UUID(data_key="instanceId")
+
+
+class APIInstanceResponseSchema(BaseSchema):
+    project_id = fields.UUID(data_key="projectId")
+    api_id = fields.UUID(data_key="apiId")
+    error_code = fields.String(data_key="errorCode")
+    error = fields.String()
+    remaining_resource_limits = fields.Nested(
+        RemainingResourceLimitSchema, data_key="remainingResourceLimits"
+    )
+
+
+class APIDevInstanceResponseSchema(APIInstanceResponseSchema):
+    instance = fields.Nested(DevInstanceIdSchema)
+    key = fields.Nested(_APIKeySchema)
+    instance_id = fields.UUID(data_key="instanceId")
+
+
+class InstanceSizeSchema(BaseSchema):
+    milli_cpus = fields.Integer(data_key="milliCpus", required=True)
+    memory_mb = fields.Integer(data_key="memoryMb", required=True)
+
+
+class APIType(str, Enum):
+    WSGI = "wsgi"
+    SCRIPT = "script"
+    PLUMBER = "plumber"
+
+
+class DeploymentStatus(Enum):
+    NOTDEPLOYED = "not-deployed"
+    STARTING = "starting"
+    DEPLOYED = "deployed"
+    ERROR = "error"
+
+
+class InstanceSchema(BaseSchema):
+    instance = fields.Nested(ServerSchema, required=True)
+    outdated = fields.Boolean(required=True)
+    key = fields.Nested(
+        _APIKeySchema, load_default=None
+    )  # Production instances do not have this field
+
+    @post_load
+    def make_instance(self, data, **kwargs):
+        return Instance(**data)
+
+
+class CommandDefinitionSchema(BaseSchema):
+    api_type = EnumField(
+        APIType, by_value=True, required=True, data_key="type"
+    )
+    working_directory = fields.String(
+        data_key="workingDirectory",
+    )
+    module = fields.String()
+    wsgi_object = fields.String(data_key="wsgiObject")
+    script_name = fields.String(data_key="scriptName")
+
+    @post_load
+    def make_command_definition(self, data, **kwargs):
+        if data["api_type"] == "wsgi":
+            return WSGIDefinition(**data)
+        elif data["api_type"] == "script":
+            return ScriptDefinition(**data)
+        elif data["api_type"] == "plumber":
+            return PlumberDefinition(**data)
```

Comment:
It's a bit odd that we're setting an api_type field on these definition objects despite the fact that this information is already encoded in the type. I think what we really want is to use something like `marshmallow-oneofschema` to serialise/deserialised between the correct types - note that we have a version of `marshmallow-oneofschema` vendored in this library.

Reply:
Redone it with the oneofschema pattern, it's pretty neat :)